core/src/main/scala/Diet.scala

```diff
@@ -0,0 +1,330 @@
+/**
+ * Created by Nicolas A Perez (@anicolaspp) on 2/4/16.
+ */
+
+package dogs
+
+import dogs.Order._
+import dogs.Predef._
+import scala.annotation.tailrec
+
+/**
+  * Discrete Interval Encoding Tree (Diet).
+  * It stores subsets of types having a total order, a predecessor and a successor function described by Enum[A]
+  *
+  * Diet is a binary search tree where each node contains a range of values and the set of all nodes is a set of
+  * disjoint sets.
+  *
+  * In the best case, when there are no "holes" in the stored set, the interval representation consists of just one
+  * single interval (node) and finding, inserting and deleting operations are O(1). In the worse case, where there
+  * are not two adjacent elements in the set, the representation is equivalent to a binary search tree.
+  */
+sealed abstract class Diet[A] {
+  import Diet._
+
+  val isEmpty: Boolean
+
+  /**
+    * verify x is a value is in the tree
+    */
+  def contains(x: A)(implicit discrete: Enum[A]): Boolean = this match {
+    case EmptyDiet()          =>  false
+    case DietNode(a,b, l, r)  =>  (discrete.compare(x, a), discrete.compare(x, b)) match {
+      case (EQ, _)  =>  true
+      case (_, EQ)  =>  true
+      case (GT, LT) =>  true
+      case (LT, _)  =>  l.contains(x)
+      case (_, GT)  =>  r.contains(x)
+    }
+  }
+
+  /**
+    * return a list of all disjoint sets in the tree where each set is represented by ARange
+    */
+  def disjointSets(): List[ARange[A]] = this match {
```

Comment:
not side-effecting, please remove the parameter list:

def disjointSets: List[ARange[A]]...